File: js/src/CopticCalendar.ts
import { computeEaster, addDays } from './algorithms.js'
import { GregorianDate } from './GregorianDate.js'
import { CopticDate } from './CopticDate.js'
import { Feast } from './Feast.js'
import { FEASTS } from './feasts-data.js'

export class CopticCalendar {
  private constructor() {}

  static easterDate(gregorianYear: number): GregorianDate {
    const [y, m, d] = computeEaster(gregorianYear)
    return new GregorianDate(y, m, d)
  }

  static moveableFeast(feastId: string, gregorianYear: number): Feast {
    const data = FEASTS.find(f => f.id === feastId && f.type === 'moveable')
    if (!data) throw new Error(`Unknown moveable feast: ${feastId}`)
    const easter = this.easterDate(gregorianYear)
    const [y, m, d] = addDays(easter.year, easter.month, easter.day, data.easter_offset!)
    const gDate = new GregorianDate(y, m, d)
    return new Feast(data, gDate, gDate.toCoptic())
  }

  static fixedFeasts(gregorianYear: number): Feast[] {
    // A Gregorian year spans two Coptic years — check both
    const cYearStart = new GregorianDate(gregorianYear, 1, 1).toCoptic().year
    const result: Feast[] = []
    const seen = new Set<string>()
    for (const copticYear of [cYearStart, cYearStart + 1]) {
      for (const data of FEASTS.filter(f => f.type === 'fixed')) {
        try {
          const c = new CopticDate(copticYear, data.coptic_month!, data.coptic_day!)
          const g = c.toGregorian()
          if (g.year === gregorianYear && !seen.has(data.id)) {
            seen.add(data.id)
            result.push(new Feast(data, g, c))
          }
        } catch { /* invalid date for this coptic year, skip */ }
      }
    }
    return result
  }

  static yearFeasts(gregorianYear: number): Feast[] {
    const moveable = FEASTS
      .filter(f => f.type === 'moveable')
      .map(data => this.moveableFeast(data.id, gregorianYear))
    const fixed = this.fixedFeasts(gregorianYear)
    return [...fixed, ...moveable].sort((a, b) => {
      const ag = a.gregorianDate, bg = b.gregorianDate
      return new Date(ag.year, ag.month - 1, ag.day).getTime()
           - new Date(bg.year, bg.month - 1, bg.day).getTime()
    })
  }
}

File: dart/lib/src/coptic_calendar.dart
import 'algorithms.dart';
import 'coptic_months_data.dart';
import 'errors.dart';
import 'feast.dart';
import 'feasts_data.dart';
import 'gregorian_date.dart';

/// Entry points for Easter and feast lookups.
class CopticCalendar {
  CopticCalendar._(); // prevent instantiation

  /// Return the Coptic month name for [month] (1..13) in [locale].
  ///
  /// Throws [InvalidCopticMonthException] if [month] is outside 1..13, and
  /// [UnsupportedLocaleException] if [locale] has no translation.
  static String monthName(int month, String locale) {
    if (month < 1 || month > 13) {
      throw InvalidCopticMonthException(month);
    }
    final record = kCopticMonths[month - 1];
    final name = record.names[locale];
    if (name == null) {
      throw UnsupportedLocaleException('', locale);
    }
    return name;
  }

  /// Return the Gregorian date of Coptic / Orthodox Easter for [gregorianYear].
  static GregorianDate easterDate(int gregorianYear) {
    final r = computeEaster(gregorianYear);
    return GregorianDate(r.year, r.month, r.day);
  }

  /// Resolve a moveable feast (by id) to its Gregorian date in [gregorianYear].
  /// Throws [ArgumentError] if [feastId] is unknown or refers to a fixed feast.
  static Feast moveableFeast(String feastId, int gregorianYear) {
    final rec = feastById(feastId);
    if (rec.type != 'moveable') {
      throw ArgumentError('feast "$feastId" is not moveable');
    }
    final easter = computeEaster(gregorianYear);
    final d = addDays(easter.year, easter.month, easter.day, rec.easterOffset!);
    return Feast(
      id: rec.id,
      type: rec.type,
      category: rec.category,
      names: rec.names,
      gregorianDate: GregorianDate(d.year, d.month, d.day),
    );
  }

  /// Return every feast (fixed + moveable) in [gregorianYear], sorted ascending by date.
  static List<Feast> yearFeasts(int gregorianYear) {
    final out = <Feast>[];
    for (final rec in kFeasts) {
      if (rec.type == 'fixed') {
        out.add(_fixedFeast(rec, gregorianYear));
      } else {
        out.add(moveableFeast(rec.id, gregorianYear));
      }
    }
    out.sort((a, b) {
      if (a.gregorianDate.year != b.gregorianDate.year) {
        return a.gregorianDate.year.compareTo(b.gregorianDate.year);
      }
      if (a.gregorianDate.month != b.gregorianDate.month) {
        return a.gregorianDate.month.compareTo(b.gregorianDate.month);
      }
      return a.gregorianDate.day.compareTo(b.gregorianDate.day);
    });
    return out;
  }

  /// Resolve a fixed Coptic feast to its Gregorian date inside [gregorianYear].
  ///
  /// A Coptic month/day falls in two possible Coptic years that overlap with
  /// the same Gregorian year. Try both candidates, keep the one landing
  /// inside [gregorianYear]; fall back to the earlier candidate.
  static Feast _fixedFeast(FeastRecord rec, int gregorianYear) {
    final cYearA = gregorianToCoptic(gregorianYear, 1, 1).year;
    final cYearB = gregorianToCoptic(gregorianYear, 12, 31).year;
    final candidates = <({int year, int month, int day})>[];
    final seen = <int>{};
    for (final cy in [cYearA, cYearB]) {
      if (!seen.add(cy)) continue;
      final d = copticToGregorian(cy, rec.copticMonth!, rec.copticDay!);
      if (d.year == gregorianYear) candidates.add(d);
    }
    if (candidates.isEmpty) {
      candidates.add(copticToGregorian(cYearA, rec.copticMonth!, rec.copticDay!));
    }
    candidates.sort((a, b) {
      if (a.year != b.year) return a.year.compareTo(b.year);
      if (a.month != b.month) return a.month.compareTo(b.month);
      return a.day.compareTo(b.day);
    });
    final d = candidates.first;
    return Feast(
      id: rec.id,
      type: rec.type,
      category: rec.category,
      names: rec.names,
      gregorianDate: GregorianDate(d.year, d.month, d.day),
    );
  }
}
